lookout/style/format/benchmarks/expected_vnodes_number.py

```diff
@@ -0,0 +1,142 @@
+"""Util to workaround https://github.com/src-d/style-analyzer/issues/557."""
+import csv
+import functools
+import logging
+import logging.handlers
+import os
+import subprocess
+import sys
+import tempfile
+from typing import Optional
+
+from lookout.core.test_helpers import server
+from tqdm import tqdm
+
+from lookout.style.format.analyzer import FormatAnalyzer
+from lookout.style.format.benchmarks.top_repos_quality import AnalyzerContextManager, \
+    ensure_repo, handle_input_arg
+from lookout.style.format.feature_extractor import FeatureExtractor
+
+analyzer_class = FormatAnalyzer
+
+
+def _stop_bblfshd():
+    subprocess.check_call("docker rm -f expected_vnodes_bblfshd".split(),
+                          stdout=sys.stderr, stderr=sys.stderr)
+
+
+def _restart_bblfshd(first: bool=False) -> str:
+    log = logging.getLogger("expected_vnodes_number")
+    port = server.find_port()
+    log.info("Restarting bblfshd")
+    if not first:
+        _stop_bblfshd()
+    subprocess.check_call(("docker run -d --rm --name expected_vnodes_bblfshd --privileged -p "
+                           "%d:9432 bblfsh/bblfshd:v2.11.0 --log-level DEBUG" % port).split(),
+                          stdout=sys.stderr, stderr=sys.stderr)
+    log.info("Installing javascript driver")
+    subprocess.check_call("docker exec expected_vnodes_bblfshd bblfshctl driver install "
+                          "javascript docker://bblfsh/javascript-driver:v1.2.0".split(),
+                          stdout=subprocess.DEVNULL, stderr=sys.stderr)
+    address = "localhost:%d" % port
+    log.info("bblfshd available at %s" % address)
+    return address
+
+
+def get_vnodes_number(repository: str, from_commit: str, to_commit: str, port: int,
+                      bblfsh: Optional[str]) -> int:
+    """
+    Generate `QualityReport` for a repository. If it fails it returns empty reports.
+
+    :param repository: URL of repository.
+    :param from_commit: Hash of commit.
+    :param to_commit: Hash of commit.
+    :param port: Port for QualityReportAnalyzer.
+    :param bblfsh: Babelfish server address to use. Specify None to use the default value.
+    :return: Reports.
+    """
+    expected_vnodes_number = -1
+
+    def capture_debug_msg(log_debug):
+        @functools.wraps(log_debug)
+        def wrapped_log(msg, *args, **kwargs):
+            if msg != FeatureExtractor.PARSED_VNODES_MSG:
+                return log_debug(msg, *args, **kwargs)
+            nonlocal expected_vnodes_number
+            if expected_vnodes_number != -1:
+                raise RuntimeError("%s should be called only one time." % log_debug.__name__)
+            log_debug(msg, *args, **kwargs)
```

Comment:
Wow damn this is too hacky. Please solve it better. We run everything in the same process, so there should not be huge difficulties. Examples:

- Override a method
- Add a callback
- Subclass and override